pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -471,3 +472,51 @@ func validateGraph(tasks []PipelineTask) *apis.FieldError {
 	}
 	return nil
 }
+
+func validateWhenSkipped(tasks []PipelineTask) (errs *apis.FieldError) {
+	d, err := dag.Build(PipelineTaskList(tasks))
+	if err != nil {
+		return apis.ErrInvalidValue(err.Error(), "tasks")
+	}
+	for i, t := range tasks {
+		if t.WhenSkipped != "" {
+			if t.WhenExpressions == nil || hasResourceDependencies(t, toMap(tasks), d) {
```

Comment:
I wonder if it might be worth having a separate error message for the `hasResourceDependencies` check?

```go
Task %q defines whenSkipped but child Task %q has a resource dependency on it
```

The reason being that I think it might be a bit confusing for the error to only say that the `whenSkipped` field is disallowed in this case without additional info.

It would be even more amazing if we could break it down further using information computed similarly to `resourceDeps()` like, `Task %q has resource dependency on Task %q (param %q references result %q)`, but that's totally not necessary for this changeset I don't think!

Reply:
Oh, also a small optimization nit: we are calling `toMap` each iteration of the for loop but I think we only need to do so once beforehand.